src/Microsoft.AspNet.OData.Shared/Builder/ODataInstanceAnnotationContainer.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// Default implementation for IODataInstanceAnnotationContainer
+    /// </summary>
+    public class ODataInstanceAnnotationContainer : IODataInstanceAnnotationContainer
+    {
+        private IDictionary<string, IDictionary<string, object>> instanceAnnotations;
+
+
+        /// <summary>
+        /// Default Constructor
+        /// </summary>
+        public ODataInstanceAnnotationContainer()
+        {
+            instanceAnnotations = new Dictionary<string, IDictionary<string, object>>();
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to the CLR type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddAnnotation(string annotationName, object value)
+        {
+            SetInstanceAnnotation(string.Empty, annotationName, value);
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to a property
+        /// </summary>
+        /// <param name="propertyName">Name of the property</param>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddPropertyAnnotation(string propertyName, string annotationName, object value)
+        {
+            SetInstanceAnnotation(propertyName, annotationName, value);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from CLR Type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <returns>Get Annotation value for the given annotation</returns>
+        public object GetAnnotation(string annotationName)
+        {
+            return GetInstanceAnnotation(string.Empty, annotationName);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from the Property
+        /// </summary>
+        /// <param name="propertyName">Name of the Property</param>
+        /// <param name="annotationName">Name of the Annotation</param>
+        /// <returns>Get Annotation value for the given annotation and property</returns>
+        public object GetPropertyAnnotation(string propertyName, string annotationName)
+        {
+            return GetInstanceAnnotation(propertyName, annotationName);
+        }
+
+        /// <summary>
+        /// Get All Annotations from CLR Type
+        /// </summary>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllAnnotations()
```

Comment:
>GetAllAnnotations [](start = 43, length = 17)

I wonder if we should call the annotations on the object something like "Get/Set[All]ObjectAnnotation[s]" (or Get/Set[All]TypeAnnotation[s]) to distinguish between *object* and *property* annotations.  I could go either way, but "GetAllAnnotations" sounds like it should include property annotations, which is not our intention for the method.